cql3/restrictions/partition_key_restrictions.cc

```diff
@@ -52,5 +53,70 @@ const expr::single_column_restrictions_map& partition_key_restrictions::get_sing
 bool partition_key_restrictions::partition_key_restrictions_is_empty() const {
     return is_empty_restriction(_partition_restrictions);
 }
+
+static bool contains_column(const expression& e) {
+    return find_in_expression<column_value>(e, [](const column_value&) { return true; }) != nullptr;
+}
+
+void partition_key_restrictions::analyze_restrictions() {
+    std::map<const column_definition*, std::vector<expression>> column_eq_restrictions;
+    std::vector<expression> token_range_restrictions;
+    std::vector<expression> filtering_restrictions;
+
+    for_each_boolean_factor(_partition_restrictions, [&](const expression& e) {
+        const binary_operator* binop = as_if<binary_operator>(&e);
+        if (binop == nullptr) {
+            filtering_restrictions.push_back(e);
+            return;
+        }
+
+        if (auto lhs_col = as_if<column_value>(&binop->lhs)) {
+            if (lhs_col->col->is_partition_key() && (binop->op == oper_t::EQ || binop->op == oper_t::IN) &&
+                !contains_column(binop->rhs)) {
+                column_eq_restrictions[lhs_col->col].push_back(e);
+                return;
+            }
+        }
+
+        if (auto lhs_token = as_if<token>(&binop->lhs)) {
+            if ((binop->op == oper_t::EQ || binop->op == oper_t::IN || is_slice(binop->op)) &&
+                !contains_column(binop->rhs)) {
+                token_range_restrictions.push_back(e);
+                return;
+            }
+        }
+
+        filtering_restrictions.push_back(e);
+    });
+
+    for (auto& [col_def, col_def_eq_restrictions] : column_eq_restrictions) {
+        _column_eq_restrictions[col_def] = conjunction{std::move(col_def_eq_restrictions)};
```

Comment:
These should be set if and only if these restrictions are useful (all columns are set).

If some column is specified more than once, we should just pick one instance and move the other to filtering.

Unless the goal here is to match existing practice.

Reply:
Filling this map is a neat way to check if every column has some eq restrictions.
I feel like doing it in a different way would be more convoluted.
To only check if each column has some eq restrictions, we would have to create a `set<const column_definition*>`, walk through the boolean factors and add columns to this set. Then we would create the map anyway. It's easier to just create the map and then check if it covers all columns or not.

> If some column is specified more than once, we should just pick one instance and move the other to filtering.

The current code supports queries like:
```
cqlsh:ks> CREATE TABLE t (pk int primary key);
cqlsh:ks> SELECT * FROM t WHERE pk = 0 AND pk = 1;
```
without any filtering.

If we kept `pk = 0`, but moved `pk = 1` that would require `ALLOW FILTERING` and cause a regression.
`possible_lhs_values` deals with such cases easily, so it's best to put all such restrictions in a single expression.

Reply:
ok.